doc/cluster_design.md

```diff
@@ -0,0 +1,137 @@
+# Design: Clustering in SQLflow to analyze patterns in data
+
+## Concept
+
+For analysts and real business people, in the daily analysis work, most of the work is not prediction, but analysis of the patterns in the data. This can help them mine user behavioral characteristics and differences, helping the business discover value and operate.
+
+This design doc introduces how to support the `Cluster Model` in SQLFlow. 
+
+## User interface
+
+In this scenario, we focus on the extraction of data patterns in unsupervised learning. SO, we plan to use a **TRAIN SQL** to train a unsupervised model. We will support whether to perform Pretrain at the beginning of this unsupervised network in `WITH`, and whether to use the already trained model as a pre-training in `USING`. The simple pipeline like:
+
+
+TRAIN SQL:
+
+``` sql
+SELECT * FROM input_table
+TRAIN clusterModel
+WITH
+    model.encode_units = [100, 7]
+    model.n_clusters = 5
+    model.run_pretrain = false
+COLUMN m1, m2, m3, m4, m5, m6, m7, m8, m9, m10 
+USING model.existed_pretrain_model =  existed_pretrain_model
+INTO my_cluster_model;
+```
+
+PREDICT SQL:
+
+``` sql
+SELECT *
+FROM input_table
+PREDICT output_table
+USING my_cluster_model;
+```
+
+where:
+- `input_table` is the high-dimensional table to be clustered.
+- `model.encode_units` is the autoencoder model layer's encoder units, the decode_units can reverse encode_units directly.
+- `model.n_clusters` is the number of patterns after clustering.
+- `my_cluster_model` is the trained cluster model.
+- `run_pretrain`  is used to determine if autoencoder pretrain needs to be run, default true.
+- `model.existed_pretrain_model` is used to specify an existing pretrain_model
+- `output_table` is the cluster result for input_table, which is adding the `group_id` column predicted by the cluster model to the input_table.
+
+## clusterModel Details
+<img src="figures/cluster_model_train_overview.png">
+
+The below figure demonstrates overall workflow for clusterModel train. This figure includes two parts, the pretrian autoencode model and the cluster model are included.
+1. First, the former is used to train a pretrain model. The `model.encode_units` describes the layer structure of the encoder of the autoencoder network. We only use the output of the trained encode layer (10000*7) as the input to the clustering model. 
+2. Then, the clustering model starts training, randomly initializes weights and multiple iterations, generates clustering models.
+3. Finally, the overall train process ultimately outputs an unsupervised clustering model.
```

Comment:
How about split `Cluster` section into `Train` and `Predict` so that users can know what does `TRAIN SQL` and `PREDICT SQL` do.